all-sdk/autotuning.md

```diff
@@ -0,0 +1,452 @@
+
+Additionally, there is a maximum cache size for cached workflows. In Core: `max_cached_workflows`. This is also set by
+users at the moment, and autotuning it is subject to the same lack of knowledge we have as with slot tuning.
+
+## Constraints
+
+There are some invariants that must be maintained:
+
+* The current number of active polls for a task type must be <= the number of free slots (in use or not) for that task
+  type at all times. Otherwise, we can end up receiving a task that won't have anywhere to go until a slot frees. For
+  example, in Core this is expressed by using a permit reservation system.
+* Max outstanding workflow tasks must be <= max cached workflows.
+
+## Proposed interface for slot management
+
+To reiterate, we can't possibly know as well as the user what resources a given workflow or activity task will consume.
+Thus, we need to provide an interface by which users can tell us whether they think it's reasonable to allow the use
+of an additional slot.
+
+The interface definition here is provided in Rust, but need not be copied exactly by other languages. Since users are
+implementing it, we want something idiomatic for them. Rust will likely immediately convert all newly allowed slots into
+a permit struct which will automatically release the slot when dropped (like it does today, without this interface). 
+Languages with destructor semantics should do the same, as permit/slot leaking is easy to do by accident.
+
+There is some mild generic magic going on to represent different types of slots and constrain that the right info type
+is provided. These guarantees can't be compile-time enforced across the language boundary, but are at least useful after
+runtime validation.
+
+```rust
+trait SlotSupplier {
+  type SlotKind: SlotKind;
+  /// Blocks until a slot is available. In languages with explicit cancel mechanisms, this should be cancellable and
+  /// return a boolean indicating whether a slot was actually obtained or not. In Rust, the future can simply
+  /// be dropped if the reservation is no longer desired.
+  async fn reserve_slot(&self, ctx: &dyn SlotReservationContext<Self::SlotKind>);
+
+  /// Tries to immediately reserve a slot, returning true if a slot is available.
+  fn try_reserve_slot(&self, ctx: &dyn SlotReservationContext<Self::SlotKind>) -> bool;
+
+  /// Marks a slot as actually now being used. This is separate from reserving one because the pollers need to
+  /// reserve a slot before they have actually obtained work from server. Once that task is obtained (and validated)
+  /// then the slot can actually be used to work on the task.
+  /// 
+  /// Users' implementation of this can choose to emit metrics, or otherwise leverage the information provided by the
+  /// `info` parameter to be better able to make future decisions about whether a slot should be handed out.
+  fn mark_slot_used(&self, info: <Self::SlotKind as SlotKind>::Info<'_>);
```

Comment:
Hrmm, yeah I first thought the "neat example" blocked in this call, but I see it doesn't. It means slot reservation may be unnecessarily expensive for things that never actually use the slot. Basically it turns remote semaphores to be based on poll instead of usage. But this is just a product of how we obtain work from server. We should at least document that people should not block in "mark used" since we don't give them async to do so.

Reply:
I feel pretty strongly that marking used can't ever block at all (and yeah I can mention that in the comment).

The reasoning is that it _happened_, and the implementer has no choice but to accept that fact. If that then means, later, that they take a long time to hand out the next permit that's acceptable, but they can't block on mark used because a) it gums things up, and b) they can't change anything anyway, it's too late.

Reply:
Works for me (can consider this resolved but can leave discussion open for others if you want)